shared/datastore.go

```diff
@@ -12,9 +16,40 @@ import (
 	"google.golang.org/appengine/datastore"
 )
 
+// Key abstracts an int64 based datastore.Key
+type Key interface {
+	IntID() int64
+}
+
+// Iterator abstracts a datastore.Iterator
+type Iterator interface {
+	Next(dst interface{}) (Key, error)
+}
+
+// Query abstracts a datastore.Query
+type Query interface {
+	Filter(filterStr string, value interface{}) Query
+	Project(project string) Query
+	Limit(limit int) Query
+	Offset(offset int) Query
+	Order(order string) Query
+	KeysOnly() Query
+	Distinct() Query
+	Run(Datastore) Iterator
+}
+
+// Datastore abstracts a datastore, hiding the distinctions between cloud and
+// appengine's datastores.
+type Datastore interface {
+	Context() context.Context
```

Comment:
Regarding Context, if we want to include it in the abstracted API, then we should make it clear that this interface should not be reused and should be constructed in each request.

Reply:
Well, for the cloud case, it's independent of requests, so I'm not sure it matters (and a context object is created per-request for the `NewAppEngineDatastore` param in any case).